src/main.rs

```diff
@@ -1,35 +1,690 @@
+/// that instance is the empty type". If you wanted to assign a gift value to that variable, you
+/// would then only be allowed to put in an empty type instance `empty_box = GiftBox::Gifts(())`.
+/// You could not change the type for `empty_box`. For example, the following would not compile:
+/// ```
+/// let mut empty_box: GiftBox<()> = GiftBox::Empty;
+/// empty_box = GiftBox::Gifts(true);
+/// // ^^^This would not compile!^^^
+///
+/// let mut int_box = GiftBox::Gifts(42);
+/// num_box = GiftBox::Gifts("Words");
+/// // ^^^This would not compile as well!^^^
+/// ```
+/// The compiler must always know what `T` is for every variable. If the value is a concrete
+/// instance of `GiftBox` then it must know the type for `T`.
+///
+/// #Generic Functions and `GiftBox`
+/// If you would like to accept a GiftBox, but this particular function does not care what kind of
+/// gift is in the gift box, then you can make the function itself generic. Example:
+/// ```
+/// fn open_gift_or_panic<T>(gift: GiftBox<T>) -> T {
+///      return match gift {
+///         GiftBox::Gifts(the_gift) => the_gift,
+///         GiftBox::Empty => panic!("I thought I was going to get a gift! D: "),
+///     }
+/// }
+/// ```
+/// Here, the function works for any `T` because the code works regardless of what the type in the
+/// gift is, it only needs to return the type which is whatever type was in the gift box. In the
+/// case of the generic function, it passes the responsibility of figuring out what the `T` is on to
+/// the caller of the function.
+///
+#[derive(Copy, Debug, PartialEq, PartialOrd)]
+pub enum GiftBox<T> {
     Gifts(T),
     Empty
 }
 
-/// In this example an empty gift box is declared and no gift box is declared with gifts. This is
-/// doomed to compilation failure because the compiler does not know what type the gifts could be.
-fn main() {
-    // Declare an empty gift box called `empty_box`.
-    let empty_box = GiftBox::Empty;
+impl<T> GiftBox<T> {
+    /// The `fill(t: Option<T>)` method accepts an [`std::option::Option`] and returns a `GiftBox`
+    /// with either `GiftBox::Gifts(t)` if `Some(t)` was provided or a `GiftBox::Empty` if
+    /// `None` is provided.
+    ///
+    /// #Arguments
+    /// * `t: Option<T>` - `t` accepts an [`std::option::Option`]
+    ///
+    /// #Returns
+    /// Returns a `GiftBox<T>`.
+    ///
+    /// #Example
+    /// Filling a `GiftBox` with `Some(T)`:
+    /// ```
+    /// let filled_box = GiftBox::fill(Some(["Toys", "Candy", "Money"]));
+    /// assert_eq!(filled_box, GiftBox::Gifts(["Toys", "Candy", "Money"]));
+    /// ```
+    /// Filling a `GiftBox` with `None`:
+    /// ```
+    /// let empty_box: GiftBox<()> = GiftBox::fill(None);
+    ///  assert_eq!(empty_box, GiftBox::Empty);
+    /// ```
+    pub fn fill(t: Option<T>) -> GiftBox<T> {
+        match t {
+            Some(t) => GiftBox::Gifts(t),
+            None => GiftBox::Empty,
+        }
+    }
```

Comment:
This functions purpose is to be able to form a gift box out of an option. Instead of implementing a custom function for that I would implement an available trait from the standard library to make the code more portable. In this case, you could implement the https://doc.rust-lang.org/std/convert/trait.From.html trait.

This is a standard trait that inicates that the thing implementing it can be created from another value. The implementation would be outside this `impl<T> GiftBox<T>` block, and would be a separate block below this:

```rust
impl<GiftBoxType> From<Option<GiftBoxType>> for GiftBox<GiftBoxType> {
    // Copied from definition of from trait
    fn from(maybeGift: Option<GiftBoxType>) -> Self {
        match t {
            Some(t) => GiftBox::Gifts(t),
            None => GiftBox::Empty,
        }
    }
}
```

This lets the "from" function be used in the exact same way your fill function is currently used, but also allows code to specify "I accept anything that a gift box can be made from". There is another trait https://doc.rust-lang.org/std/convert/trait.Into.html (from/into are somwhat interchangable and imply each other) which is automatically implemented when you implement from, and we can use the Into trait as a type constraint in a function:

```rust
fn box_up_gift<GiftType, GiftSource: Into<GiftType>>(source: GiftSource) -> GiftBox<GiftType> {
    source.into()
}
```

That way, if some other type could be made into a gift box this would accept that as a GiftSource type as well. If you later decided you wanted to implement From<OtherType> (and thus automatically Into) for something other than Option<T> then this function would automatically accept it, since it doesnt really carewhat the GiftSource is so long as it can be made into a giftbox.

Because of the fact that implementing the From trait automatically implements the Into trait, and we are using the Into trait as our type constraint, this example might not be the most clear, but hopefully its not too confusing. The TLDR is that if a behavior you are going for is something that might be reusable (such as creating an X from a Y) then its worth checking if there is an existing trait that you can implement, or creating a new trait to represent that behavior for your own reuse in the future